### PR TITLE
Improve `MessageMapping` rules dispatch

### PR DESCRIPTION
**Pull request by bvdmitri**
This PR improves upon #290 and addresses #284 even further. 

This PR changes the way `MessageMapping` dispatches on a particular rule. I couldn't improve things further because it requires a lot of changes in the core, but there is definitely room for improvement and we can reduce allocation profile even further. This is very crucial for real-time applications.

The PR is breaking, but imo minor. After this PR, it is no longer possible to add/change/override built-in rules without explicit import statement, e.g:
```
import ReactiveMP: rule_for_NormalMeanVariance # `@rule` below will error without this import statement

@rule NormalMeanVariance(:out, ...) ... = begin 
 ...
end
```
There are no problems adding new rules with the explicit import statement. I could think more of possible solutions for that, but in general further improvements require a complete `rule` refactoring.

The benefits are improved allocation profile and faster inference in general, especially for the `rxinference` function from `RxInfer`. These are timing from the `Tiny Benchmark.ipynb`. 

### `master`

```
BenchmarkTools.Trial: 1834 samples with 1 evaluation.
 Range (min … max):  2.457 ms … 11.236 ms  ┊ GC (min … max): 0.00% … 74.35%
 Time  (median):     2.518 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.725 ms ±  1.177 ms  ┊ GC (mean ± σ):  6.67% ± 11.31%

  █▅▁                                                         
  ███▆▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅▆▇ █
  2.46 ms      Histogram: log(frequency) by time     10.1 ms <

 Memory estimate: 3.05 MiB, allocs estimate: 83828.
BenchmarkTools.Trial: 565 samples with 1 evaluation.
 Range (min … max):  7.731 ms … 21.617 ms  ┊ GC (min … max): 0.00% … 54.27%
 Time  (median):     7.899 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   8.850 ms ±  2.809 ms  ┊ GC (mean ± σ):  9.13% ± 14.68%

  █▆▂▂                                                        
  █████▆▅▆▆▅▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▄▆▆▁▄▅▆▇▆▅▇▅▄ ▆
  7.73 ms      Histogram: log(frequency) by time     18.8 ms <

 Memory estimate: 11.10 MiB, allocs estimate: 241913.
```

### `PR`

```
BenchmarkTools.Trial: 2831 samples with 1 evaluation.
 Range (min … max):  1.576 ms …   9.150 ms  ┊ GC (min … max): 0.00% … 79.29%
 Time  (median):     1.615 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.765 ms ± 998.268 μs  ┊ GC (mean ± σ):  8.03% ± 11.32%

  █▂                                                           
  ██▅▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ █
  1.58 ms      Histogram: log(frequency) by time      8.65 ms <

 Memory estimate: 2.56 MiB, allocs estimate: 67828.

BenchmarkTools.Trial: 603 samples with 1 evaluation.
 Range (min … max):  6.953 ms … 21.804 ms  ┊ GC (min … max):  0.00% … 55.96%
 Time  (median):     7.264 ms              ┊ GC (median):     0.00%
 Time  (mean ± σ):   8.295 ms ±  3.085 ms  ┊ GC (mean ± σ):  10.30% ± 15.45%

  █▆▅▃▂▁ ▁                                                    
  ████████▆▅▆▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▁▁▄▄▆▇██▆▆▅▅▁▄▄ ▇
  6.95 ms      Histogram: log(frequency) by time     19.8 ms <

 Memory estimate: 10.73 MiB, allocs estimate: 229906.
```

WDYT of the benefits and of the breaking change? 

@bartvanerp you might be interested.

### CODE REVIEW

**Comment by bvdmitri** [comment]
This is the README benchmark with this PR
image

**Comment by bvdmitri** [comment]
Benchmarks:

<details><summary>Memory allocation profile</summary>
<p>

```
benchmark_ar_03-03-2023-17-31_v1.8.5.txt: Memory estimate: 121.36 MiB, allocs estimate: 2216360.Julia Version 1.8.5
benchmark_ar_03-03-2023-17-44_v1.8.5.txt: Memory estimate: 111.34 MiB, allocs estimate: 1887996.Julia Version 1.8.5
benchmark_cvi_03-03-2023-17-39_v1.8.5.txt: Memory estimate: 4.01 GiB, allocs estimate: 130044341.Julia Version 1.8.5
benchmark_cvi_03-03-2023-17-53_v1.8.5.txt: Memory estimate: 4.01 GiB, allocs estimate: 129867541.Julia Version 1.8.5
benchmark_gmm_multivariate_03-03-2023-17-33_v1.8.5.txt: Memory estimate: 239.39 MiB, allocs estimate: 3403106.Julia Version 1.8.5
benchmark_gmm_multivariate_03-03-2023-17-46_v1.8.5.txt: Memory estimate: 239.39 MiB, allocs estimate: 3403085.Julia Version 1.8.5
benchmark_gmm_univariate_03-03-2023-17-32_v1.8.5.txt: Memory estimate: 14.53 MiB, allocs estimate: 295558.Julia Version 1.8.5
benchmark_gmm_univariate_03-03-2023-17-45_v1.8.5.txt: Memory estimate: 14.53 MiB, allocs estimate: 295544.Julia Version 1.8.5
benchmark_hgf_03-03-2023-17-36_v1.8.5.txt: Memory estimate: 268.21 MiB, allocs estimate: 6464076.Julia Version 1.8.5
benchmark_hgf_03-03-2023-17-49_v1.8.5.txt: Memory estimate: 244.10 MiB, allocs estimate: 5804067.Julia Version 1.8.5
benchmark_iid_mv_covariance_03-03-2023-17-38_v1.8.5.txt: Memory estimate: 75.98 MiB, allocs estimate: 1432147.Julia Version 1.8.5
benchmark_iid_mv_covariance_03-03-2023-17-51_v1.8.5.txt: Memory estimate: 69.80 MiB, allocs estimate: 1252140.Julia Version 1.8.5
benchmark_iid_mv_covariance_known_mean_03-03-2023-17-38_v1.8.5.txt: Memory estimate: 45.68 MiB, allocs estimate: 861086.Julia Version 1.8.5
benchmark_iid_mv_covariance_known_mean_03-03-2023-17-51_v1.8.5.txt: Memory estimate: 42.93 MiB, allocs estimate: 771086.Julia Version 1.8.5
benchmark_iid_mv_precision_03-03-2023-17-36_v1.8.5.txt: Memory estimate: 70.99 MiB, allocs estimate: 1357129.Julia Version 1.8.5
benchmark_iid_mv_precision_03-03-2023-17-50_v1.8.5.txt: Memory estimate: 64.58 MiB, allocs estimate: 1177122.Julia Version 1.8.5
benchmark_iid_mv_wishart_known_mean_03-03-2023-17-37_v1.8.5.txt: Memory estimate: 41.10 MiB, allocs estimate: 801087.Julia Version 1.8.5
benchmark_iid_mv_wishart_known_mean_03-03-2023-17-50_v1.8.5.txt: Memory estimate: 38.35 MiB, allocs estimate: 711087.Julia Version 1.8.5
benchmark_lar_03-03-2023-17-31_v1.8.5.txt: Memory estimate: 275.21 MiB, allocs estimate: 2406969.Julia Version 1.8.5
benchmark_lar_03-03-2023-17-44_v1.8.5.txt: Memory estimate: 271.43 MiB, allocs estimate: 2301948.Julia Version 1.8.5
benchmark_linreg_03-03-2023-17-40_v1.8.5.txt: Memory estimate: 19.18 MiB, allocs estimate: 507196.Julia Version 1.8.5
benchmark_linreg_03-03-2023-17-54_v1.8.5.txt: Memory estimate: 16.66 MiB, allocs estimate: 429675.Julia Version 1.8.5
benchmark_linreg_broadcasted_03-03-2023-17-40_v1.8.5.txt: Memory estimate: 19.12 MiB, allocs estimate: 506004.Julia Version 1.8.5
benchmark_linreg_broadcasted_03-03-2023-17-54_v1.8.5.txt: Memory estimate: 16.61 MiB, allocs estimate: 428483.Julia Version 1.8.5
benchmark_mlgssm_03-03-2023-17-34_v1.8.5.txt: Memory estimate: 63.09 MiB, allocs estimate: 1279598.Julia Version 1.8.5
benchmark_mlgssm_03-03-2023-17-35_v1.8.5.txt: Memory estimate: 21.58 MiB, allocs estimate: 400093.Julia Version 1.8.5
benchmark_mlgssm_03-03-2023-17-47_v1.8.5.txt: Memory estimate: 62.48 MiB, allocs estimate: 1259591.Julia Version 1.8.5
benchmark_mlgssm_03-03-2023-17-48_v1.8.5.txt: Memory estimate: 21.58 MiB, allocs estimate: 400093.Julia Version 1.8.5
benchmark_probit_03-03-2023-17-35_v1.8.5.txt: Memory estimate: 8.82 MiB, allocs estimate: 119065.Julia Version 1.8.5
benchmark_probit_03-03-2023-17-48_v1.8.5.txt: Memory estimate: 8.62 MiB, allocs estimate: 112722.Julia Version 1.8.5
benchmark_ulgssm_03-03-2023-17-33_v1.8.5.txt: Memory estimate: 11.10 MiB, allocs estimate: 241919.Julia Version 1.8.5
benchmark_ulgssm_03-03-2023-17-47_v1.8.5.txt: Memory estimate: 10.73 MiB, allocs estimate: 229912.Julia Version 1.8.5
```

</p>
</details>

<details><summary>Absolute timings</summary>
<p>

```
benchmark_ar_03-03-2023-17-31_v1.8.5.txt: Time  (mean ± σ):   116.498 ms ±   7.260 ms  ┊ GC (mean ± σ):  11.82% ±  6.08%
benchmark_ar_03-03-2023-17-44_v1.8.5.txt: Time  (mean ± σ):    97.747 ms ±   8.715 ms  ┊ GC (mean ± σ):  13.49% ±  8.01%
benchmark_cvi_03-03-2023-17-39_v1.8.5.txt: Time  (mean ± σ):   4.141 s ± 10.358 ms  ┊ GC (mean ± σ):  8.39% ± 0.11%
benchmark_cvi_03-03-2023-17-53_v1.8.5.txt: Time  (mean ± σ):   4.320 s ± 13.756 ms  ┊ GC (mean ± σ):  8.10% ± 0.11%
benchmark_gmm_multivariate_03-03-2023-17-33_v1.8.5.txt: Time  (mean ± σ):   160.139 ms ±   9.743 ms  ┊ GC (mean ± σ):  16.33% ±  4.21%
benchmark_gmm_multivariate_03-03-2023-17-46_v1.8.5.txt: Time  (mean ± σ):   162.557 ms ±   9.470 ms  ┊ GC (mean ± σ):  16.10% ±  4.15%
benchmark_gmm_univariate_03-03-2023-17-32_v1.8.5.txt: Time  (mean ± σ):   10.193 ms ±  3.603 ms  ┊ GC (mean ± σ):  11.17% ± 16.29%
benchmark_gmm_univariate_03-03-2023-17-45_v1.8.5.txt: Time  (mean ± σ):   10.299 ms ±  3.544 ms  ┊ GC (mean ± σ):  10.83% ± 16.06%
benchmark_hgf_03-03-2023-17-36_v1.8.5.txt: Time  (mean ± σ):   297.283 ms ±   6.101 ms  ┊ GC (mean ± σ):  7.44% ± 1.95%
benchmark_hgf_03-03-2023-17-49_v1.8.5.txt: Time  (mean ± σ):   262.137 ms ±   6.548 ms  ┊ GC (mean ± σ):  7.67% ±  2.34%
benchmark_iid_mv_covariance_03-03-2023-17-38_v1.8.5.txt: Time  (mean ± σ):   64.010 ms ±  8.502 ms  ┊ GC (mean ± σ):  11.03% ± 10.82%
benchmark_iid_mv_covariance_03-03-2023-17-51_v1.8.5.txt: Time  (mean ± σ):   52.816 ms ±  8.289 ms  ┊ GC (mean ± σ):  12.04% ± 12.24%
benchmark_iid_mv_covariance_known_mean_03-03-2023-17-38_v1.8.5.txt: Time  (mean ± σ):   35.902 ms ±  6.648 ms  ┊ GC (mean ± σ):  10.76% ± 13.48%
benchmark_iid_mv_covariance_known_mean_03-03-2023-17-51_v1.8.5.txt: Time  (mean ± σ):   31.156 ms ±  6.353 ms  ┊ GC (mean ± σ):  11.31% ± 14.30%
benchmark_iid_mv_precision_03-03-2023-17-36_v1.8.5.txt: Time  (mean ± σ):   55.125 ms ±  8.223 ms  ┊ GC (mean ± σ):  11.80% ± 11.82%
benchmark_iid_mv_precision_03-03-2023-17-50_v1.8.5.txt: Time  (mean ± σ):   44.401 ms ±  8.105 ms  ┊ GC (mean ± σ):  13.23% ± 13.71%
benchmark_iid_mv_wishart_known_mean_03-03-2023-17-37_v1.8.5.txt: Time  (mean ± σ):   29.353 ms ±  6.439 ms  ┊ GC (mean ± σ):  11.75% ± 14.78%
benchmark_iid_mv_wishart_known_mean_03-03-2023-17-50_v1.8.5.txt: Time  (mean ± σ):   24.697 ms ±  6.084 ms  ┊ GC (mean ± σ):  12.84% ± 16.17%
benchmark_lar_03-03-2023-17-31_v1.8.5.txt: Time  (mean ± σ):   200.288 ms ±   8.619 ms  ┊ GC (mean ± σ):  14.01% ±  3.61%
benchmark_lar_03-03-2023-17-44_v1.8.5.txt: Time  (mean ± σ):   195.370 ms ±   8.386 ms  ┊ GC (mean ± σ):  14.25% ±  3.67%
benchmark_linreg_03-03-2023-17-40_v1.8.5.txt: Time  (mean ± σ):   19.905 ms ±  5.429 ms  ┊ GC (mean ± σ):  7.92% ± 13.81%
benchmark_linreg_03-03-2023-17-54_v1.8.5.txt: Time  (mean ± σ):   14.772 ms ±  4.994 ms  ┊ GC (mean ± σ):  9.20% ± 14.90%
benchmark_linreg_broadcasted_03-03-2023-17-40_v1.8.5.txt: Time  (mean ± σ):   19.604 ms ±  5.505 ms  ┊ GC (mean ± σ):  8.18% ± 14.06%
benchmark_linreg_broadcasted_03-03-2023-17-54_v1.8.5.txt: Time  (mean ± σ):   14.432 ms ±  4.963 ms  ┊ GC (mean ± σ):  9.32% ± 14.99%
benchmark_mlgssm_03-03-2023-17-34_v1.8.5.txt: Time  (mean ± σ):   67.363 ms ±  11.348 ms  ┊ GC (mean ± σ):  11.72% ± 12.99%
benchmark_mlgssm_03-03-2023-17-35_v1.8.5.txt: Time  (mean ± σ):   18.625 ms ±  4.669 ms  ┊ GC (mean ± σ):  9.37% ± 14.42%
benchmark_mlgssm_03-03-2023-17-47_v1.8.5.txt: Time  (mean ± σ):   66.131 ms ± 11.018 ms  ┊ GC (mean ± σ):  11.70% ± 13.03%
benchmark_mlgssm_03-03-2023-17-48_v1.8.5.txt: Time  (mean ± σ):   18.956 ms ±  4.749 ms  ┊ GC (mean ± σ):  9.32% ± 14.40%
benchmark_probit_03-03-2023-17-35_v1.8.5.txt: Time  (mean ± σ):   9.242 ms ±  2.548 ms  ┊ GC (mean ± σ):  6.56% ± 12.66%
benchmark_probit_03-03-2023-17-48_v1.8.5.txt: Time  (mean ± σ):   8.785 ms ±  2.481 ms  ┊ GC (mean ± σ):  6.64% ± 12.70%
benchmark_ulgssm_03-03-2023-17-33_v1.8.5.txt: Time  (mean ± σ):   9.283 ms ±  3.404 ms  ┊ GC (mean ± σ):  9.75% ± 15.12%
benchmark_ulgssm_03-03-2023-17-47_v1.8.5.txt: Time  (mean ± σ):   8.590 ms ±  3.342 ms  ┊ GC (mean ± σ):  10.22% ± 15.43%
```

</p>
</details>

TLDR: Memory allocation has improved for all benchmarks. Timings have improved almost for all examples, except for mixtures and CVI where speed remained around the same.

**Comment by bartvanerp** [comment]
I think this is great work @bvdmitri! 

The performance gain is actually huge in my opinion, but the breaking change is something we should have a better looks at because it conflicts with user-friendliness. Just a (probably stupid) suggestion, but could the @rule macro expand to include the import statement automatically?

**Comment by bvdmitri** [comment]
You are right @bartvanerp . We, unfortunately, cannot add import statement automatically because by default there is no way to tell in which module the node has been defined (it is not necessarily `ReactiveMP`). Perhaps, I can save that information to use for import though. 

I shall try it out or maybe I can find another way to implement it.

**Comment by bvdmitri** [comment]
@bartvanerp I played around a little bit and I think I found even a better approach that does not have any breaking changes and actually gives even a little bit more improvement (in terms of allocations). Please leave this PR open for now and do not approve it.
